alphadia/workflow/peptidecentric.py

```diff
@@ -319,121 +282,198 @@ def get_batch_plan(self):
+                ms1_optimizer = optimization.AutomaticMS1Optimizer(
+                    self.config["search_initial"]["initial_ms1_tolerance"],
+                    self,
+                )
+        else:
+            ms1_optimizer = None
         if self.dia_data.has_mobility:
-            if (
-                self.com.mobility_error
-                > self.config["search"]["target_mobility_tolerance"]
-            ):
-                self.reporter.log_string(
-                    f"❌ {'mobility_error':<15}: {self.com.mobility_error:.4f} > {self.config['search']['target_mobility_tolerance']}",
-                    verbosity="info",
+            if self.config["search"]["target_mobility_tolerance"] > 0:
+                mobility_optimizer = optimization.TargetedMobilityOptimizer(
+                    self.config["search_initial"]["initial_mobility_tolerance"],
+                    self.config["search"]["target_mobility_tolerance"],
+                    self,
                 )
-                continue_calibration = True
             else:
-                self.reporter.log_string(
-                    f"✅ {'mobility_error':<15}: {self.com.mobility_error:.4f} <= {self.config['search']['target_mobility_tolerance']}",
-                    verbosity="info",
+                mobility_optimizer = optimization.AutomaticMobilityOptimizer(
+                    self.config["search_initial"]["initial_mobility_tolerance"],
+                    self.calibration_manager,
+                    self.optimization_manager,
+                    self.fdr_manager,
                 )
-
-        if self.com.current_epoch < self.config["calibration"]["min_epochs"] - 1:
-            self.reporter.log_string(
-                f"❌ {'current_epoch':<15}: {self.com.current_epoch} < {self.config['calibration']['min_epochs']}",
-                verbosity="info",
-            )
-            continue_calibration = True
         else:
-            self.reporter.log_string(
-                f"✅ {'current_epoch':<15}: {self.com.current_epoch} >= {self.config['calibration']['min_epochs']}",
-                verbosity="info",
-            )
+            mobility_optimizer = None
 
-        self.reporter.log_string(
-            "==============================================", verbosity="info"
+        optimizers = [
+            ms2_optimizer,
+            rt_optimizer,
+            ms1_optimizer,
+            mobility_optimizer,
+        ]
+        targeted_optimizers = [
+            [
+                optimizer
+                for optimizer in optimizers
+                if isinstance(optimizer, optimization.TargetedOptimizer)
+            ]
+        ]
+        automatic_optimizers = [
+            [optimizer]
+            for optimizer in optimizers
+            if isinstance(optimizer, optimization.AutomaticOptimizer)
+        ]
+
+        order_of_optimization = (
+            targeted_optimizers + automatic_optimizers
+            if any(targeted_optimizers)
```

Comment:
It's not redundant: `any(targeted_optimizers) ` returns False for `targeted_optimizers=[[]]`, which is the case if there are no targeted optimizers in use (because of the way the optimizer lists are nested so all targeted optimizers run together). If the conditional is not present, this gives a result of [[], [automatic_optimizer_1], [automatic_optimizer_2], ...] in which the empty list causes an infinite loop later.

However, this isn't very clear from the code. I should at least write a comment to explain it, or think of a way to rewrite it.